### PR TITLE
Add Parallel option for apply-patch

### PR DESCRIPTION
**Pull request by xincunli-sonic**
#### What I did

Add parallel option for apply-patch if the device is multi asic.

#### How I did it

Use multithread to handle each asic config apply.

#### How to verify it
1. Single ASIC

```shell
admin@str3-s6100-acs-6:~/gcu$ cat empty.json 
[]

admin@str3-s6100-acs-6:~/gcu$ sudo config apply-patch -p empty.json 
Patch Applier: localhost: Patch application starting.
Patch Applier: localhost: Patch: []
Patch Applier: localhost getting current config db.
Patch Applier: localhost: simulating the target full config after applying the patch.
Patch Applier: localhost: validating all JsonPatch operations are permitted on the specified fields
Patch Applier: localhost: validating target config does not have empty tables,
                            since they do not show up in ConfigDb.
Patch Applier: localhost: sorting patch updates.
Patch Applier: The localhost patch was converted into 0 changes.
Patch Applier: localhost: applying 0 changes in order.
Patch Applier: localhost: verifying patch updates are reflected on ConfigDB.
Patch Applier: localhost patch application completed.
Patch applied successfully.

admin@str3-s6100-acs-6:~/gcu$ sudo config apply-patch empty.json 
Patch Applier: localhost: Patch application starting.
Patch Applier: localhost: Patch: []
Patch Applier: localhost getting current config db.
Patch Applier: localhost: simulating the target full config after applying the patch.
Patch Applier: localhost: validating all JsonPatch operations are permitted on the specified fields
Patch Applier: localhost: validating target config does not have empty tables,
                            since they do not show up in ConfigDb.
Patch Applier: localhost: sorting patch updates.
Patch Applier: The localhost patch was converted into 0 changes.
Patch Applier: localhost: applying 0 changes in order.
Patch Applier: localhost: verifying patch updates are reflected on ConfigDB.
Patch Applier: localhost patch application completed.
Patch applied successfully.
```

2. Multi ASIC

```shell
stli@str2-7250-lc1-2:~/gcu$ cat empty.json 
[]

stli@str2-7250-lc1-2:~/gcu$ sudo config apply-patch -p empty.json 
sonic_yang(6):Note: Below table(s) have no YANG models: DHCP_SERVER
Patch Applier: localhost: Patch application starting.
Patch Applier: localhost: Patch: []
Patch Applier: localhost getting current config db.
Patch Applier: asic0: Patch application starting.
Patch Applier: asic1: Patch application starting.
Patch Applier: asic0: Patch: []
Patch Applier: asic1: Patch: []
Patch Applier: asic0 getting current config db.
Patch Applier: asic1 getting current config db.
Patch Applier: localhost: simulating the target full config after applying the patch.
Patch Applier: localhost: validating all JsonPatch operations are permitted on the specified fields
Patch Applier: localhost: validating target config does not have empty tables,
                            since they do not show up in ConfigDb.
Patch Applier: localhost: sorting patch updates.
Patch Applier: asic0: simulating the target full config after applying the patch.
Patch Applier: asic0: validating all JsonPatch operations are permitted on the specified fields
Patch Applier: asic0: validating target config does not have empty tables,
                            since they do not show up in ConfigDb.Patch Applier: asic1: simulating the target full config after applying the patch.

Patch Applier: asic1: validating all JsonPatch operations are permitted on the specified fields
Patch Applier: asic0: sorting patch updates.
Patch Applier: asic1: validating target config does not have empty tables,
                            since they do not show up in ConfigDb.
Patch Applier: asic1: sorting patch updates.
Patch Applier: The localhost patch was converted into 0 changes.
Patch Applier: localhost: applying 0 changes in order.
Patch Applier: localhost: verifying patch updates are reflected on ConfigDB.
Patch Applier: The asic0 patch was converted into 0 changes.
Patch Applier: asic0: applying 0 changes in order.
Patch Applier: asic0: verifying patch updates are reflected on ConfigDB.
Patch Applier: localhost patch application completed.
Patch Applier: The asic1 patch was converted into 0 changes.
Patch Applier: asic1: applying 0 changes in order.
Patch Applier: asic1: verifying patch updates are reflected on ConfigDB.
Patch Applier: asic0 patch application completed.
Patch Applier: asic1 patch application completed.
Patch applied successfully.

stli@str2-7250-lc1-2:~/gcu$ sudo config apply-patch empty.json 
sonic_yang(6):Note: Below table(s) have no YANG models: DHCP_SERVER
Patch Applier: localhost: Patch application starting.
Patch Applier: localhost: Patch: []
Patch Applier: localhost getting current config db.
Patch Applier: localhost: simulating the target full config after applying the patch.
Patch Applier: localhost: validating all JsonPatch operations are permitted on the specified fields
Patch Applier: localhost: validating target config does not have empty tables,
                            since they do not show up in ConfigDb.
Patch Applier: localhost: sorting patch updates.
Patch Applier: The localhost patch was converted into 0 changes.
Patch Applier: localhost: applying 0 changes in order.
Patch Applier: localhost: verifying patch updates are reflected on ConfigDB.
Patch Applier: localhost patch application completed.
Task ('', []): ended on thread 140442129442624
Patch Applier: asic0: Patch application starting.
Patch Applier: asic0: Patch: []
Patch Applier: asic0 getting current config db.
Patch Applier: asic0: simulating the target full config after applying the patch.
Patch Applier: asic0: validating all JsonPatch operations are permitted on the specified fields
Patch Applier: asic0: validating target config does not have empty tables,
                            since they do not show up in ConfigDb.
Patch Applier: asic0: sorting patch updates.
Patch Applier: The asic0 patch was converted into 0 changes.
Patch Applier: asic0: applying 0 changes in order.
Patch Applier: asic0: verifying patch updates are reflected on ConfigDB.
Patch Applier: asic0 patch application completed.
Task ('asic0', []): ended on thread 140442129442624
Patch Applier: asic1: Patch application starting.
Patch Applier: asic1: Patch: []
Patch Applier: asic1 getting current config db.
Patch Applier: asic1: simulating the target full config after applying the patch.
Patch Applier: asic1: validating all JsonPatch operations are permitted on the specified fields
Patch Applier: asic1: validating target config does not have empty tables,
                            since they do not show up in ConfigDb.
Patch Applier: asic1: sorting patch updates.
Patch Applier: The asic1 patch was converted into 0 changes.
Patch Applier: asic1: applying 0 changes in order.
Patch Applier: asic1: verifying patch updates are reflected on ConfigDB.
Patch Applier: asic1 patch application completed.
Task ('asic1', []): ended on thread 140442129442624
Patch applied successfully.
```

### CODE REVIEW

**Review by wen587 — DISMISSED** [review]
lgtm

File: tests/config_test.py [hunk 1]
```diff
@@ -2965,6 +2965,77 @@ def test_apply_patch_multiasic(self):
...
+                # Mock ConfigDBConnector to ensure it's not called during dry-run
+                with patch('config.main.ConfigDBConnector') as mock_config_db_connector:
+
+                    print("Multi ASIC: {}".format(multi_asic.is_multi_asic()))
+                    # Invocation of the command with the CliRunner
+                    result = self.runner.invoke(config.config.commands["apply-patch"],
+                                                [self.patch_file_path,
+                                                 "--format", ConfigFormat.SONICYANG.name,
+                                                 "--dry-run",
+                                                 "--parallel",
```
Comment (qiluo-msft) [inline tests/config_test.py:3018]: >parallel [](http://example.com/codeflow?start=52&length=8)

How do you test it actually operate in parallel?

Comment (xincunli-sonic) [reply tests/config_test.py:3018]: added check for multiple thread called if parallel option set.

File: tests/config_test.py [hunk 2]
```diff
@@ -2965,6 +2965,156 @@ def test_apply_patch_multiasic(self):
...
+
+                # Mock ConfigDBConnector to ensure it's not called during dry-run
+                with patch('config.main.ConfigDBConnector') as mock_config_db_connector:
+
+                    print("Multi ASIC: {}".format(multi_asic.is_multi_asic()))
+                    # Invocation of the command with the CliRunner
+                    result = self.runner.invoke(config.config.commands["apply-patch"],
+                                                [self.patch_file_path,
+                                                 "--format", ConfigFormat.SONICYANG.name,
+                                                 "--dry-run",
```
Comment (wen587) [inline tests/config_test.py:2982]: Hi @xincunli-sonic , I saw these tests are only run with `dry-run` operation. is there a way to check it work as expected without `dry-run`?

File: tests/config_test.py [hunk 3]
```diff
@@ -2965,6 +2965,154 @@ def test_apply_patch_multiasic(self):
...
+                    self.assertIn("Patch applied successfully.", result.output)
+
+                    # Assertions to check if ThreadPoolExecutor was used correctly
+                    MockThreadPoolWait.assert_called_once()
+
+                    # Verify mocked_open was called as expected
+                    mocked_open.assert_called_with(self.patch_file_path, 'r')
+
+                    # Ensure ConfigDBConnector was never instantiated or called
+                    mock_config_db_connector.assert_not_called()
```
Comment (wen587) [inline tests/config_test.py:3074]: Is it correct without dry-run? The config should be updated by patch. Why it is not being called?

File: tests/config_test.py (hunk 1 above)
Comment (qiluo-msft) [reply tests/config_test.py:3018]: I see your new testcase `test_apply_patch_check_running_in_not_parallel_multiasic`. However this comment is still applicable.

**Review by wen587 — APPROVED** [review]
lgtm

**Review by wen587 — DISMISSED** [review]
lgtm

**Review by qiluo-msft — APPROVED** [review]

**Pull request merged**